src/Features/Core/Portable/CodeFixes/Configuration/ConfigurationUpdater.cs

```diff
@@ -397,23 +398,104 @@ private AnalyzerConfigDocument FindOrGenerateEditorConfig()
-                    mostRecentHeader = groups[1].Value.ToString().ToLowerInvariant();
+                    var mostRecentHeaderText = groups[1].Value.ToString().ToLowerInvariant();
+
+                    if (mostRecentHeaderText.Equals("*"))
+                    {
+                        lastValidHeader = mostRecentHeader;
+                    }
+                    else
+                    {
+                        var brokenUpHeader = mostRecentHeaderText.Split(new[] { '.' }, 2);
+                        var brokenUpFileExtensions = brokenUpHeader[1].Split(',', ' ', '{', '}');
+
+                        // Replacing characters in the header with the regex equivalent
+                        brokenUpHeader[0] = brokenUpHeader[0].Replace("*", ".*");
+                        brokenUpHeader[0] = brokenUpHeader[0].Replace("/", @"\/");
+
+                        // Creating the header regex string, ex. [*.{cs,vb}] => ((\.cs)|(\.vb))
+                        var headerRegexStr = brokenUpHeader[0] + @"((\." + brokenUpFileExtensions[0] + ")";
+                        for (var i = 1; i < brokenUpFileExtensions.Length; i++)
+                        {
+                            headerRegexStr += @"|(\." + brokenUpFileExtensions[i] + ")";
+                        }
+                        headerRegexStr += ")";
+
+                        var headerRegex = new Regex(headerRegexStr);
+
+                        // We check that the relative path of the .editorconfig file to the diagnostic file
+                        // matches the header regex pattern.
+                        if (headerRegex.IsMatch(relativePath))
+                        {
+                            var match = headerRegex.Match(relativePath).Value.Split('.');
+
+                            // Edge case: The below statement checks that we correctly handle cases such as a header of [m.cs] and
+                            // a file name of Program.cs.
+                            if (match[0].Contains(PathUtilities.GetFileName(diagnosticFilePath, false)))
+                            {
+                                // If the diagnostic's isPerLanguage = true, the rule is valid for both C# and VB.
+                                // For the purpose of adding missing rules later, we want to keep track of whether there is a
+                                // valid header that contains both [*.cs] and [*.vb]. 
+                                // If isPerLanguage = false or a compiler diagnostic, the rule is only valid for one of the languages.
+                                // Thus, we want to keep track of whether there is an existing header that only contains [*.cs] or only
+                                // [*.vb], depending on the language.
+                                // We also keep track of the last valid header for the language.
+                                if (!SuppressionHelpers.IsCompilerDiagnostic(_diagnostic) && _isPerLanguage &&
```

Comment:
probably can extract out `!SuppressionHelpers.IsCompilerDiagnostic(_diagnostic) && _isPerLanguage` into a separate local and the below else if can then just use `!local_name`?